proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,527 @@
+    # Team names can contain upper and lower case characters,
+    # whitespace, and punctuation (e.g. "Kube admins!").
+    # Slug names are lower case alphanumeric characters and may
+    # contain dashes (e.g. "kube-admins").
+    # Either way, group names will always be prefixed by the
+    # GitHub org name followed by a colon (e.g. my-org:my-team).
+    # Note that colons are not allowed in GitHub org names,
+    # so the first colon in the group name will always be the
+    # org/team separator colon.
+    # If desired, an admin could configure identity
+    # transformation expressions on a FederationDomain to
+    # further customize these group names.
+    # Defaults to "slug".
+    groups: slug
+  # Optionally reject any user who attempts to authenticate
+  # unless they belong to one of these GitHub orgs.
+  # The GitHub App or GitHub OAuth App provided in the
+  # spec.client.secretName must be allowed by the org
+  # owners to view org memberships, or else the
+  # user will not be considered to belong to that org
+  # or belong to any teams within that org.
+  # When specified, users' group memberships will be
+  # filtered to include only those GitHub teams which
+  # are owned by these orgs.
+  # If desired, an admin could configure identity
+  # policy expressions on a FederationDomain to
+  # further customize which users and groups are allowed
+  # to authenticate.
+  # When not set, any GitHub user can authenticate
+  # regardless of org membership, and all the team
+  # memberships returned by the GitHub API for that
+  # user will be reflected as groups regardless of
+  # which orgs own those teams.
+  allowOrganizations: [ vmware-tanzu, broadcom ]
+  # Allow or disallow users to use GitHub Personal Access Tokens
+  # for CLI-based (no web browser) authentication.
+  # Setting these both to false disables all non-interactive
+  # CLI-based auth, but still allows browser-based auth.
+  allowPersonalAccessTokens:
+    # Allow or disallow the use of GitHub Fine-grained PATs
+    # to authenticate to Kubernetes clusters. If they
+    # are disallowed for your GitHub org, then also
+    # disallow them here so your users can get a nice
+    # error message when trying to use a fine-grained PAT.
+    # Defaults to false.
+    fineGrained: true
+    # Allow or disallow the use of GitHub Classic PATs
+    # to authenticate to Kubernetes clusters. If they
+    # are disallowed for your GitHub org, then also
+    # disallow them here so your users can get a nice
+    # error message when trying to use a classic PAT.
+    # Defaults to false.
+    classic: false
+  client:
+    # The name of Secret in the same namespace which holds the
+    # client ID and client secret of a GitHub App or
+    # Github OAuth App. This will only be used for web
+    # browser-based auth flows (not CLI-based flows).
+    # Required.
+    secretName: github-client-credentials
+status:
+  conditions: # conditions TBD. Shows validations, if any.
```

Comment:
We should enumerate some of the expected status conditions. If necessary, we can wait for the implementation phase so that we are aware of what Pinniped admins would want to see, but these should be enumerated in some design doc.

Examples of known server-side validations:
- `client.secretName` must be a valid secret in the same namespace.
- Any validations on `allowOrganizations` and `allowUsersInAllGitHubOrgs` that need to take place (this behavior has not been finalized)
- `github_api.certificateAuthorityData` should be base64-encoded if provided (can we validate that it's a PEM-encoded certificate?)
- `github_api.host` is reachable and TLS verification succeeds. We could also require that the user prefix this with the `https://` schema, or simply assume `https://`.

Reply:
TODO: Add this to proposal.

Reply:
Agree, worth reviewing existing Conditions for consistency, then add anything new (such as 1 condition per spec field).

Reply:
I was just looking at LDAP and OIDC.  These seem a little light on first pass, maybe with GitHub we can provide a more robust set of status conditions, then circle back later and improve the older ones.

Reply:
> Verify that the list of provided allowedOrganizations are actual organizations

I think we decided that this is not desired because it is too hard to get a token to act as a client itself. It would require the Supervisor admin to do lots of extra configuration in GitHub and Pinniped, which is not desirable.

Reply:
The `.Status.Condition[]` set in a bit more detail could look something like this:

```yaml
- `GitHubIdentityProvider`
  - `GitHubIdentityProviderSpec` `json:"spec"`
   - `GitHubAPIConfig`
     - `Host`
       - Has `CEL` validations, but older versions of Kubernetes will not run these.
         - Therefore we should also do validations in the controller
       - `url.Parse`, these should be IPv4,IPv6,hostnames.
         - We can allow a port, using the same address helpers as WebhookAuthenticator, default to :443
         - do not specify https://, similar to ldapadp or ADadp.
           - CEL will fail if https is added, so controller should also.
     - `TLS`
       - Same as Concierge TLSSpec
         - CertificateAuthorityData
           - if specified, we should try a dial again.
           - if equal to default values (github.com, nil TLS), skip the dial.
             - default is github.com (host) with no provided CA (TLSSpec empty). 
   - `GitHubClaims`
     - `Username`
       - dont need to validate the enum {"id","login","login:id"} as this works on all versions of kube
       - dont validate default=login:id, this works on all versions of kubernetes
       - so no validation at all.
     - `Groups`
       - no validation at all, again like above, CRD validations all work for old kube versions
   - `AllowedOrganizations`
     -  length of this dictates what `OrganizationLoginPolicy` MUST be
       - empty, then `OrganizationLoginPolicy` MUST be `AllOrganizationsAllowed`,
   - `OrganizationLoginPolicy`
     - Interacts with the OrganizationLoginPolicy.
       - Don't really need?
       - Include to FORCE user to say "I KNOW I AM DOING THIS"
         - AllowedOrganizationsOnly;AllOrganizationsAllowed
         - wording is confusing, can we switch?
     - If empty, then the `AllowedOrganizations` must have a value.
     - If has value `AllOrganizationsAllowed`, then AllowedOrganizations must be empty
     - If `AllowedOrganizationsOnly`, then AllowedOrganiztations must have a list
   - `Client`
     - `secretName` this just references a secret
       - This secret will need to be validated itself.
       - copy the validations from the `OIDCIdentityProviderController`
         - Does it exist?
         - Is it the right type? (slightly different secret type)
         - Does it have the right keys?
         - Are the values of the keys empty?
         - should be the same validations as this other controller

```

Reply:
Distilled down, this may be what the `.Status.Conditions[]` are specifically:

```yaml
- HostURLValid
- TLSConfigurationValid
- GitHubClaimsValid // no need to validate in controller, CRD validations sufficient
- AllowedOrganizationsValid // may or may not need, if the following validation is sufficient?
- OrganizationLoginPolicyValid // valid/invalid based on AllowedOrganizations
- ClientValid // secretRef
```

Reply:
It seems like this is more like:

- `HostValid`. This is not a URL, so it must not contain `://`. It must be parsable as defined by our `endpointaddr.Parse()` helper.
- `TLSConfigurationValid`. The CA bundle can be parsed as a CA bundle.
- `AllowedOrganizationsPolicyValid`. Use the same logic and same error text as the CEL validations, in case an old version of Kubernetes is being used.
- `ClientCredentialsValid`. This is what we call the same check in the OIDCIdentityProvider's status.
- Something about successfully dialing. In LDAPIdentityProvider we called this `LDAPConnectionValid`.

For reference, the helper mentioned above is https://github.com/vmware-tanzu/pinniped/blob/d9a3992b3b6445ea37ebd693b40405b719079f18/internal/endpointaddr/endpointaddr.go#L32-L41

TODO: update the proposal to list these status checks